[WIP] World Model refactor and new features

orduno opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info |  |
| ------ | ----------- |
| Ticket(s) this addresses   |  #379  |
| Primary OS tested on | Ubuntu 16.04 |
| Robotic platform tested on | Gazeobo simulation of turtlebot |

---

## Description of contribution

* Refactored `nav2_world_model`. Defined an interface for specifying different world representations (`WorldRepresentation`) and moved current functionality to a `CostmapRepresentation`.
* Added service for checking if a region is free of obstacles. 
* Added a `RegionVisualizer` for displaying regions in RVIZ that get processed by the world model.
* Defined a World Model client class.
* Modified the Global Planner `Navfn` and  the `Backup` motion primitive to use the new world model client interface.
* Added a new method in the `MotionPrimitive` class called `pathIsClear()` that motion primitives can implement for checking if there is enough space for execution.
* Modified `Backup` so it checks if there is enough space for backing up before executing.
* Added a new nav2 service description for specifying a rectangular region.
* Added a `robot_centric_costmap` with only the obstacle layer enabled and centered on the robot. 

---

## Future work

* Obtain robot footprint on motion primitives #524 *

crdelsey commented:
@orduno Can you fix the linter failures before we start reviewing?

SteveMacenski commented on nav2_dwb_controller/dwb_controller/include/dwb_controller/dwb_controller.hpp:
```diff
@@ -22,10 +22,13 @@
 namespace nav2_dwb_controller
 {
 
+enum class TravelDirection {MovingForward, MovingBackwards, Stopped};
```
Can this be expanded and made explicit? ie
```
enum TravelDirection
{
  MovingForward = 1
  Moving...
}

```

SteveMacenski replied:
The intent here was actually to expand the enum out so its not a 1 liner that's easily glossed over (also not sure the enum class is needed, just enum is sufficient?). My general preference is to remove it, I think it's not state information that needs to be stored. Explicitly defining the numbers is just habit, I suppose it doesn't add value

SteveMacenski commented on nav2_dwb_controller/dwb_controller/include/dwb_controller/dwb_controller.hpp:
```diff
@@ -47,6 +51,10 @@ class DwbController : public rclcpp::Node
+
+  // Proxy to the World Model
+  nav2_world_model::WorldModelClient world_model_;
+  TravelDirection travel_direction_;
```
The `travel_direction_` seems like unnecessary state. The return code for `checkRegion()` isnt used so the function doesn't seem to have much value. Also the DWB planner will handle the potential velocity commands and compute valid ones or come back with a failure code. Not sure what this is accomplishing

orduno replied:
Yes, the return code of `checkRegion()` is not used and probably won't be used by DWB. I mainly added that to test the client interface and dynamically visualize some region (in the direction the robot is traveling).

![2](https://user-images.githubusercontent.com/39749557/50924657-13238e00-1405-11e9-8df2-3cde17e194e4.gif)

`travel_direction_` helps in specifying if we should look ahead or backwards

![3](https://user-images.githubusercontent.com/39749557/50924836-862d0480-1405-11e9-8888-15835c177940.gif)

I'm ok with either removing entirely this on DWB or adding some parameter for the user to enable or disable. If we keep it, we should consider the overhead it adds.

 @crdelsey Do you have a preference on this?

SteveMacenski replied:
It's a useful feature, I think though it belongs in a higher level location and uses the clear information to do something (i.e. limit velocity, send a stop command, etc, configurably)

crdelsey replied:
@orduno  It's misleading in DWB. It suggests the robot is looking at that region, when that is not how DWB works. DWB only looks at the costmap below the generated trajectories, and it looks both behind and in front, and to the side as well potentially.

SteveMacenski replied:
This is a useful thing to setup boundaries for collision (stop) zones and slow down zones, its super common for autonomous fork lifts and tape following robots to utilize shapes like that to check if it needs to stop. It's a higher level thing, maybe belonging to the robot class implementation of whether to use and the configurations of those shape parameters (and whether to use costmap and/or direct laser scans)

orduno replied:
Ok, I'll remove it from DWB.

SteveMacenski commented on nav2_dwb_controller/dwb_controller/src/dwb_controller.cpp:
```diff
@@ -70,6 +71,7 @@ DwbController::followPath(const nav2_tasks::FollowPathCommand::SharedPtr command
         RCLCPP_INFO(get_logger(), "No pose. Stopping robot");
         publishZeroVelocity();
       } else {
+        checkRegion(pose2d);
```
The return code isnt used and DWB compute velocity commands will look at the correct time projection for collisions

orduno replied:
Agree, please review my comment above.

SteveMacenski commented on nav2_motion_primitives/src/back_up.cpp:
```diff
@@ -108,4 +126,55 @@ nav2_tasks::TaskStatus BackUp::controlledBackup()
+      //              has not localized and initial pose has not been received.
+      //              In this case we might want to directly check laser scanner measurements
+      //              for possible obstacles.
+      RCLCPP_WARN(node_->get_logger(), "Backup primitive is not checking for free space."
```
This is a motion primitive for multiple users, I dont think there should be any reference to auto-localization which is an independent and up stream item. It breaks the encapsulation of making a general backup maneuver .

orduno replied:
Agree, will work on a solution for this.

orduno replied:
I have removed the check for initial robot pose altoghether.

We'll enable a service on the world model that checks for free space relative to the robot using directly the laser scanner. That way even if the robot is not localized, world model can still provide obstacle information around it. Motion primities, since are executing short movements, can use this service.

As you can see, I've created a new branch in origin called `world_model_refactor`, once I finish addressing the feedback on this PR. I'll work on a second PR that creates this new service.

SteveMacenski commented on nav2_motion_primitives/src/back_up.cpp:
```diff
@@ -108,4 +126,55 @@ nav2_tasks::TaskStatus BackUp::controlledBackup()
   return TaskStatus::FAILED;
 }
 
+bool BackUp::pathIsClear()
```
I still don't fully understand this API - I agree that there's probably a need in most motions to check for collision, but this API doesnt give it a set of waypoints, or a time to project forward, or any important information to make that decision based on. This function in backup doesn't actually produce a `return true` and requests the pose that somewhere else in the call was probably determined to pass along

orduno replied:
I assume you are talking about the `WorldModel` API, not the `MotionPrimitive` base class. In the case of the API for checking for free space I tried to consider the simplest useful case, i.e. check within some rectangular region and return false if any cell is occupied, in this case it's up to the client to define the region. We can later extend the API with a more sofisticated approach like providing waypoints to the World Model (instead of the description of the rectangular region).

> This function in backup doesn't actually produce a return true and requests the pose that somewhere else in the call was probably determined to pass along

Could you clarify the second part of the question?

SteveMacenski commented on nav2_motion_primitives/src/spin.cpp:
```diff
@@ -146,4 +146,10 @@ nav2_tasks::TaskStatus Spin::controlledSpin()
   return TaskStatus::RUNNING;
 }
 
+bool Spin::pathIsClear()
```
See https://github.com/ros-planning/navigation/blob/kinetic-devel/rotate_recovery/src/rotate_recovery.cpp#L105

No need to reinvent the wheel :)

orduno replied:
Yeah, the code in `controlledSpin()` is based on ROS1 `rotate_recovery` controller. I haven't added a service for checking for free space within the robot's own [footprint](https://github.com/ros-planning/navigation/blob/kinetic-devel/rotate_recovery/src/rotate_recovery.cpp#L122).

SteveMacenski commented on nav2_motion_primitives/src/stop.cpp:
```diff
@@ -70,4 +70,9 @@ nav2_tasks::TaskStatus Stop::onCycleUpdate(nav2_tasks::StopResult & /*result*/)
   return nav2_tasks::TaskStatus::SUCCEEDED;
 }
 
+bool Stop::pathIsClear()
```
This is a good instance where this method doesn't necessarily make sense. I think the API should just have an `IsValid()` method to return true if so or false when it couldnt do something necessary. It should be up to the primative to check if the path is clear because if it wasn't, it shouldnt be able to compute a valid trajectory/set of actions

orduno replied:
I think this requires more discussion. I'm assuming all motion primitives produce some robot movement and requires to check if the region of operation is clear. If that's the case, we should make it explicit on the API, that's why I have the `pathIsClear` method on the `MotionPrimitive` base class.

Now about the stop primitive, I don't think it should be a primitive but instead a method on the robot class.

orduno replied:
It's not that straight forward, it would require some changes to bt_navigator. Will follow up with #575.

SteveMacenski commented on nav2_navfn_planner/package.xml:
```diff
@@ -16,6 +16,7 @@
   <build_depend>nav2_util</build_depend>
   <build_depend>nav2_msgs</build_depend>
   <build_depend>nav2_robot</build_depend>
+  <build_depend>nav2_world_model</build_depend>
```
Thanks :)

SteveMacenski commented on nav2_navfn_planner/src/navfn_planner.cpp:
```diff
@@ -478,11 +479,12 @@ NavfnPlanner::getCostmap(
-  auto request = std::make_shared<nav2_tasks::CostmapServiceClient::CostmapServiceRequest>();
-  request->specs.resolution = 1.0;
+  nav2_world_model::CostmapServiceRequest request;
+  request.specs.resolution = 1.0;
```
better default of 0.05?

orduno replied:
True

SteveMacenski commented on nav2_world_model/include/nav2_world_model/costmap_representation.hpp:
```diff
@@ -0,0 +1,74 @@
+    rclcpp::executor::Executor & executor,
+    rclcpp::Clock::SharedPtr & clock);
+
+  GetCostmap::Response getCostmap(const GetCostmap::Request & request) override;
```
Might want to think of a better API name than `getCostmap` as sometimes people might not be using costmaps (ie elevation maps, non-regular-grid-maps, etc)

orduno replied:
Certainly, different world model clients migh require to obtain different expressions of the world. I think we will be improving this API as we go.

However, we should point out that the internal world representation used should be independent of the interface. If world model has a client, navfn for instance, that requires a costmap, the world representation should support the conversion.

crdelsey replied:
I think the API needs to be thought through with some of those other representations in mind. Is there a use case for using a elevation map as the internal representation, but then converting it to a costmap? If so, should it be the responsibility of the elevation map to make that conversion? Or should the representations provide their data, with the conversion being the responsibility of another component.

orduno replied:
All of these are valid points and have been included in #565.

This PR is a minor refactor of the world model, #565 will be a redesign of the world model. For now, map server and global planner use cost maps so no need to extend the API.

SteveMacenski commented on nav2_world_model/include/nav2_world_model/region_visualizer.hpp:
```diff
@@ -0,0 +1,134 @@
+  // ROS node used to create the publishers and subscribers
+  rclcpp::Node::SharedPtr node_;
+
+  rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr marker_publisher_;
```
You may find a marker array is easier (and expandable to use for non-rectangular shapes)

orduno replied:
I'm assuming that the region visualizer will use the same type of marker for each cell and the color will provide meaning.

We can extend it later if we see a use for sending a bunch of different shapes for visualizing a single region.

SteveMacenski commented on nav2_world_model/include/nav2_world_model/world_model_client.hpp:
```diff
@@ -0,0 +1,67 @@
+  bool clearArea(const ClearAreaServiceRequest & specs)
+  {
+    auto request = std::make_shared<ClearAreaServiceRequest>(specs);
+    return free_space_client_.invoke(request).get()->was_successful;
```
`was_successful` indicates to me that the call to the service was successful, not that the area was free, could be changed to something closer to `clear`?

orduno replied:
Thanks for pointing this out, will push a change.

SteveMacenski commented on nav2_world_model/src/costmap_representation.cpp:
```diff
@@ -0,0 +1,186 @@
+}
+
+ProcessRegion::Response
+CostmapRepresentation::clearArea(const ProcessRegion::Request & /*request*/)
```
It seems to be these clearArea, confirmFreeSpace, etc methods should sit on top of the costmap implementation, not be part of it. I'm still not entirely sure what these are doing exactly given a planner or controller should look at a costmap and make the decision about the validity of the action based on the map presented. Having something in the main navigator loop that queries a configurable shape for something in the costmap or laser scan at loop rate makes sense to add an additional level of safety, but I dont think that belongs here along side t he costmaps, but maybe in the WorldModel class or as a separate utility.

SteveMacenski replied:
Or part of this Robot API

orduno replied:
I'll try to address your comment, but this can be a lenghy topic of discussion for the workgroup meeting.

We are working towards creating a world model that is independent of the underlying world representation (currently we only have costmap, but there are others). That means we need to define what any world representation should provide to the world model.

> It seems to be these clearArea, confirmFreeSpace, etc methods should sit on top of the costmap implementation, not be part of it.

That's the goal of the `CostmapRepresentation`. It's providing the inteface required by the world model to use a costmap. A costmap should be able, among other things, to confirm if some region is occupied by calling `confirmFreeSpace`.

> I'm still not entirely sure what these are doing exactly... 

These methods define the required information for the current type of world model clients used on the stack -- (global) planner and motion primitives. The DWA costmap has not been consolidated to the world model yet.

> ... given a planner or controller should look at a costmap and make the decision about the validity of the action based on the map presented.

That's an option, providing the entire costmap to the local planner, however from a design point of view, it's better to understand what exactly is the information needed by the controller. For example, it's my understanding that the dwa controller needs the costmap to check if a possible trajectory hits an object, that could be a service on the world model, instead of having each controller implement a different algorithm for checking if a trajectory goes into an object.

> Having something in the main navigator loop that queries a configurable shape for something in the costmap or laser scan at loop rate makes sense to add an additional level of safety

Yes the `confirmFreeSpace` could be used by the safetly layer, but not exclusively. Motion primitives might, possibly some type of planner. Therefore it makes sense to have it as a general service of the world model.

> but I dont think that belongs here along side t he costmaps, but maybe in the WorldModel class or as a separate utility.

It is a service of the world model and is something the costmap should provide, the `CostmapRepresentation` is just adapting the interface of the world model to the one fo the costmap, for example:

backup_primitive --(freeSpace?)--> world_model --(freeSpace?)--> world_representation

As the stack grows to support other families of controllers, planners, primitives, world representations, multiple robots (and a shared comprehesion of the world), this re-design will be necessary.

crdelsey commented:
@orduno Two things:
1. Smaller PRs please, or make the commit history clean and show the logically separate changes.
2. Can you put together a doc or presentation on the problem you are trying to solve, and the design you'd like to see? It seems like it is something that would be helpful to discuss in the working group meeting.

crdelsey commented on nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:
```diff
@@ -55,6 +55,16 @@ namespace nav2_costmap_2d
 // convenient for storing x/y point pairs
 struct MapLocation
 {
+  MapLocation()
```
I don't see that we need this constructor anywhere. I don't think it's that much of a burden on the user to say `MapLocation foo(0,0)` instead of `MapLocation foo`

orduno replied:
Agree, will push a change.

orduno commented on nav2_motion_primitives/src/back_up.cpp:
```diff
@@ -72,30 +86,34 @@ nav2_tasks::TaskStatus BackUp::timedBackup()
+  robot_->sendVelocity(default_vel_);
+
+  // TODO(orduno): #423 fixed time
+  auto elapsed_time_ = std::chrono::system_clock::now() - start_time_;
```
OK, however this method goes away with #521.

orduno commented on nav2_motion_primitives/src/back_up.cpp:
```diff
@@ -108,4 +126,55 @@ nav2_tasks::TaskStatus BackUp::controlledBackup()
+  // Define the region size
+  // Width is set to match the robot's diameter
+  // Height is set to estimate the distance that will be traveled
+  double robot_width = 0.22;  // TODO(orduno) get from robot class
```
I'll fix this.

orduno replied:
I've made it a parameter, will also add it to nav2_params.yaml.

orduno commented on nav2_dwb_controller/dwb_controller/src/dwb_controller.cpp:
```diff
@@ -145,4 +153,32 @@ bool DwbController::getRobotPose(nav_2d_msgs::msg::Pose2DStamped & pose2d)
   return true;
 }
 
+bool DwbController::checkRegion(nav_2d_msgs::msg::Pose2DStamped & pose2d)
```
Agree, will remove. We can later add it somewhere else.

orduno commented on nav2_motion_primitives/include/nav2_motion_primitives/motion_primitive.hpp:
```diff
@@ -65,6 +71,12 @@ class MotionPrimitive
+  // the path that will be covered by the primitive is free.
+  // The method will be called at the begining of the execution of the primitive
+  // if method returns false, execution is cancelled.
+  virtual bool pathIsClear() = 0;
```
I'm not totally sure we should remove it from the base class, I'm assuming any `MotionPrimitive` produces a robot motion and should check for space. We currently have a `Stop` primitive but purely stopping the robot should just be a method on the robot class.

orduno replied:
As discussed, we'll leave it there for now.

orduno commented on nav2_msgs/srv/ProcessRegion.srv:
```diff
@@ -0,0 +1,19 @@
+# Process a rectangular region
```
Let's have an offline discussion on this there are several things to consider.

orduno commented on nav2_dwb_controller/dwb_controller/include/dwb_controller/dwb_controller.hpp:
```diff
@@ -22,10 +22,13 @@
 namespace nav2_dwb_controller
 {
 
+enum class TravelDirection {MovingForward, MovingBackwards, Stopped};
```
I've removed this from DWB.

orduno commented on nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp:
```diff
@@ -55,6 +55,16 @@ namespace nav2_costmap_2d
+
+  MapLocation(const unsigned int x, const unsigned int y)
+  : x(x), y(y)
+  {
```
changed

orduno commented on nav2_world_model/include/nav2_world_model/clear_area_service_client.hpp:
```diff
@@ -0,0 +1,42 @@
+using ClearAreaServiceResponse =
+  ServiceClient<ProcessRegion>::ResponseType;
+
+class ClearAreaServiceClient : public ServiceClient<ProcessRegion>
```
I've added comments to this services, in this case is for clearing the occupancy information of a region, similar to `clear_costmap` in nav1.

orduno replied:
Good point, I've pushed a commit that removes the placeholders for this service.

orduno commented on nav2_util/include/nav2_util/point.hpp:
```diff
@@ -0,0 +1,74 @@
+namespace nav2_util
+{
+
+struct Point
```
changed

orduno commented:
@crdelsey 

> Can you put together a doc or presentation on the problem you are trying to solve, and the design you'd like to see? It seems like it is something that would be helpful to discuss in the working group meeting.

I can put together some slides based on the discussion we had and share it on the working group.

orduno commented:
@crdelsey @SteveMacenski I think I've addressed all your feedback. Comments that have been resolved completed should be hidden, the ones still showing might require your response.

This PR will get pushed to the `world_model_refactor` branch. There we can work on addressing #517 before merging to master, this is mainly needed because auto-localization need to check for open space even when the robot is not localized.

crdelsey commented on nav2_world_model/include/nav2_world_model/world_representation.hpp:
```diff
@@ -0,0 +1,59 @@
+
+  WorldRepresentation() = delete;
+
+  virtual GetCostmap::Response getCostmap(const GetCostmap::Request & request) = 0;
```
These different functions should be abstracted from the world model in some way, so that it can be extended with arbitrary capabilities int the future. As I see how you are using it, it seems it could be a map of std::function objects. You could do something like `worldRepresentation["GetCostmap"]()` in the client code or something.

orduno replied:
I think this will change significantly once we define the layers and plugins from #565.

crdelsey commented on nav2_world_model/include/nav2_world_model/costmap_representation.hpp:
```diff
@@ -0,0 +1,74 @@
+namespace nav2_world_model
+{
+
+class CostmapRepresentation : public WorldRepresentation
```
Does this inheritance structure still make sense when the World Model consists of plugins?

orduno replied:
To be addressed in #565.

orduno commented on nav2_motion_primitives/src/back_up.cpp:
```diff
@@ -15,8 +15,14 @@
+#include <cmath>
 
 #include "nav2_motion_primitives/back_up.hpp"
+#pragma GCC diagnostic ignored "-Wpedantic"
```
Good catch, I've pushed a change.

orduno commented on nav2_motion_primitives/src/back_up.cpp:
```diff
@@ -81,16 +88,69 @@ nav2_tasks::TaskStatus BackUp::controlledBackup()
+
+  // Translate to set the edge of the region in front of the robot
+  request.offset.x = 0.0;
+  request.offset.y = robot_width / 2.0 + request.height / 2.0;
```
I've added a method to the robot class for getting its width.

orduno commented on nav2_world_model/include/nav2_world_model/region_visualizer.hpp:
```diff
@@ -0,0 +1,134 @@
+    marker_(visualization_msgs::msg::Marker())
+  {
+    marker_publisher_ = node_->create_publisher<visualization_msgs::msg::Marker>(
+      "world_model_cell", 1);
```
That's probably ok for now since only one primitive is executed at a time, and the user might want to visualize only what the current primitive is requesting. We can extend this in the future as we see a need.

orduno commented on nav2_world_model/include/nav2_world_model/region_visualizer.hpp:
```diff
@@ -0,0 +1,134 @@
+      marker_.points.size());
+
+    marker_.header.frame_id = "map";
+    // marker_.header.stamp = node_.now();
```
Should provide the current time, will fix it.

orduno commented on nav2_world_model/include/nav2_world_model/world_model.hpp:
```diff
@@ -16,36 +16,44 @@
+  // World representation
+  std::unique_ptr<WorldRepresentation> world_representation_;
+
+  // The services provided
```
If I understand correctly, the problem is how to build a world model given different services to satisfy and different representations to use. I'll follow-up offline with you to discuss some ideas.

orduno replied:
I think this would be addressed in #565 once we identify the interface between the core-plugin-client.

SteveMacenski commented on nav2_bringup/launch/nav2_params.yaml:
```diff
@@ -39,14 +40,16 @@ global_costmap:
       plugin_names: ["obstacle_layer", "inflation_layer"]
       plugin_types: ["nav2_costmap_2d::ObstacleLayer", "nav2_costmap_2d::InflationLayer"]
       global_frame: "base_link"
+      origin_x: -5.0
```
how large is this local costmap? This origin should be -size/2 for the ray casting to work. a 10m local costmap would be a little excessive

orduno replied:
I was using for testing the default size which is 10x10 m. I'll change it on the yaml so it's something between 3 to 5 m with the origin, as you mention, at -size/2.

SteveMacenski commented on nav2_bringup/launch/nav2_params.yaml:
```diff
@@ -1,6 +1,6 @@
 map_server:
   ros__parameters:
-    yaml_filename: "test_map.yaml"
+    yaml_filename: "/home/caorduno/map.yaml"
```
hardcoded path

orduno commented:
@crdelsey @SteveMacenski - I think I've addressed all your comments. 

On top of what you previously reviewed, I've added a second costmap on the world model that only has the obstacle layer and is centered on the robot, I'm calling it `robot_centric_costmap`. This is the one that will be used by the motion primitives (and recovery behaviors).

This also enables auto-localization. -- Motion primitives can now move the robot around without crashing into objects when the robot is not localized.

In terms of the world model and interface, there will be changes as we develop the full design.

Can you please take a look? and if it's ok, approve to get it merged.

crdelsey commented:
@orduno Can you rebase to remove the unrelated travis and docker file changes before we review?

orduno commented:
@crdelsey The PR is up-to-date with ros-planning:master. I changed the branch to which this PR will get merged from 'world_model_refactor' to 'master'.

crdelsey commented on nav2_motion_primitives/include/nav2_motion_primitives/motion_primitive.hpp:
```diff
@@ -91,6 +104,19 @@ class MotionPrimitive
+  std::shared_ptr<nav2_robot::Robot> robot_;
+
+  // Proxy to the World Model
+  nav2_world_model::WorldModelClient world_model_;
```
This class never uses the world_model_. Seems likely the wrong place for it.

orduno replied:
This is the base class for the motion primitives. Trying to avoid duplicating this line across all primitives.

crdelsey commented on nav2_motion_primitives/src/back_up.cpp:
```diff
@@ -81,16 +90,74 @@ nav2_tasks::TaskStatus BackUp::controlledBackup()
+  request.reference.y = 0.0;
+  request.rotation = 0.0;
+
+  // TODO(orduno) If robot is localized we might want to check relative to the map
```
Commented out code. Remove if not needed.

orduno replied:
removed

crdelsey commented on nav2_robot/include/nav2_robot/robot.hpp:
```diff
@@ -38,6 +38,7 @@ class Robot
   bool getOdometry(nav_msgs::msg::Odometry::SharedPtr & robot_odom);
   std::string getName();
   void sendVelocity(geometry_msgs::msg::Twist twist);
+  double getWidth() {return width_;}
```
This should be more like getFootprint or at least getRadius, however, I think this needs to be reconciled with the getFootprint on costmap. I'd rather you just got the footprint off the costmap until such time as we can remove the code on costmap and put a fully functional version here.

orduno replied:
I'll change it to `getRadius`. Getting the footprint from costmap will require adding a service.

orduno replied:
And costmap should really be getting the footprint from the robot class -  #563

crdelsey commented on nav2_navfn_planner/src/navfn_planner.cpp:
```diff
@@ -478,11 +479,12 @@ NavfnPlanner::getCostmap(
-  auto request = std::make_shared<nav2_tasks::CostmapServiceClient::CostmapServiceRequest>();
-  request->specs.resolution = 1.0;
+  nav2_world_model::CostmapServiceRequest request;
+  request.specs.resolution = 0.05;
```
Why the change in resolution?

orduno replied:
Using a more reasonable value. -- although it's currently ignored when servicing the request.

crdelsey commented on nav2_world_model/include/nav2_world_model/world_model.hpp:
```diff
@@ -16,36 +16,38 @@
 private:
-  void costmap_callback(
+  // World representation
+  std::map<std::string, std::unique_ptr<WorldRepresentation>> world_representations_;
```
After seeing @SteveMacenski's comments and the discussion in the WG meeting, it seems there is a need to have multiple or alternate representations. We need to figure out how that should work though. I don't think we have the interface figured out yet.

orduno replied:
Yes, there is a need for multiple representations. We can continue design related questions on #565.

orduno commented on nav2_motion_primitives/include/nav2_motion_primitives/back_up.hpp:
```diff
@@ -34,13 +34,14 @@ class BackUp : public MotionPrimitive<nav2_tasks::BackUpCommand, nav2_tasks::Bac
+  geometry_msgs::msg::Twist default_vel_;
 
-  nav_msgs::msg::Odometry::SharedPtr initial_pose_;
+  // The requested and initial poses
```
fixed

orduno replied:
Yep, fixed

orduno commented on nav2_dwb_controller/dwb_controller/include/dwb_controller/dwb_controller.hpp:
```diff
@@ -39,6 +40,7 @@ class DwbController : public rclcpp::Node
   void publishVelocity(const nav_2d_msgs::msg::Twist2DStamped & velocity);
   void publishZeroVelocity();
   bool getRobotPose(nav_2d_msgs::msg::Pose2DStamped & pose2d);
+  bool checkRegion(nav_2d_msgs::msg::Pose2DStamped & pose2d);
```
Yep, fixed

orduno commented on nav2_motion_primitives/include/nav2_motion_primitives/back_up.hpp:
```diff
@@ -34,13 +34,14 @@ class BackUp : public MotionPrimitive<nav2_tasks::BackUpCommand, nav2_tasks::Bac
-  nav_msgs::msg::Odometry::SharedPtr initial_pose_;
+  // The requested and initial poses
   double command_x_;
+  nav_msgs::msg::Odometry::SharedPtr initial_pose_;
```
Agree, fixed

orduno replied:
Yep, fixed

orduno commented on nav2_navfn_planner/src/navfn_planner.cpp:
```diff
@@ -478,11 +479,12 @@ NavfnPlanner::getCostmap(
 
-  auto result = costmap_client_.invoke(request);
-  costmap = result.get()->map;
+  RCLCPP_INFO(get_logger(), "Received costmap of size %d, %d",
```
Yep, changed to DEBUG.

orduno commented on nav2_world_model/include/nav2_world_model/clear_area_service_client.hpp:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  std::shared_ptr<ClearAreaServiceResponse>
+  wasCleared(std::shared_ptr<ClearAreaServiceRequest> & request)
```
Yeah, it's boiler plate code, but it keeps `WorldModelClient` clean.

orduno commented on nav2_costmap_2d/src/costmap_2d_ros.cpp:
```diff
@@ -457,19 +457,19 @@ bool Costmap2DROS::getRobotPose(geometry_msgs::msg::PoseStamped & global_pose) c
-
-    return false;
-  }
+  // if (current_time - global_pose.header.stamp >
```
Pushed a fix for this.

orduno commented on nav2_world_model/include/nav2_world_model/world_representation.hpp:
```diff
@@ -0,0 +1,60 @@
+  virtual GetCostmap::Response getCostmap(const GetCostmap::Request & request) = 0;
+
+  // Verify if a region is unoccupied
+  virtual ProcessRegion::Response confirmFreeSpace(const ProcessRegion::Request & request) = 0;
```
I think I answered your question offline. Besides defining the interfaces to the world model we also need to identify the interface between the world model and the representations. For now, we define these to be equal.

orduno commented on nav2_world_model/src/costmap_representation.cpp:
```diff
@@ -0,0 +1,188 @@
+    return false;
+  }
+
+  // Check if there is at least one cell not free and send to visualizer for display
```
Not sure I understand the issue, this is similar to how costmap_2d and other nav2 modules publish messages internally for visualization in RVIZ.

This is the robot's local world model, it's own point of view. Once we add more robots, each will have it's own, and then we'll have to introduce namespaces. 

At some point, it will be beneficial to have a shared global world model, but that will have a different architecture than `nav2_world_model`.

orduno commented on nav2_util/include/nav2_util/point2d.hpp:
```diff
@@ -0,0 +1,67 @@
+           " y: " << point.y;
+  }
+
+  void rotateAroundPoint(const double theta, const Point2D & reference)
```
I did some simplifications and got rid of this function.

orduno commented on nav2_msgs/srv/ProcessRegion.srv:
```diff
@@ -0,0 +1,21 @@
+float64 width
+float64 height
+
+# Translation and rotation are specified with the variables below
```
I was trying to keep the service definition general, but agree that it can be simplified given the current need. 

Will push a change.

orduno commented on nav2_world_model/include/nav2_world_model/costmap_service_client.hpp:
```diff
@@ -0,0 +1,54 @@
+ * A class that creates a ROS service for obtaining a costmap
+ */
+
+class CostmapServiceClient : public ServiceClient<GetCostmap>
```
Ok, let's keep it on `nav2_tasks` -- we can later deal with the package becoming too heterogeneous.

As for the changes to the class itself, now that I see the current need we can remove `getCostmapService()` and just directly invoke.

orduno commented on nav2_world_model/src/costmap_representation.cpp:
```diff
@@ -0,0 +1,188 @@
+  std::vector<Point2D> vertices = {
+    Point2D{left, down}, Point2D{left, top}, Point2D{right, top}, Point2D{right, down}};
+
+  // TODO(orduno) the RVIZ world frame is rotated 90 deg with respect to Gazebo.
```
Got it fixed.

orduno commented:
Closing the PR as the main features provided have been incorporated (for example, collision checking, odom/robot centric costmap, etc.) we can continue our design discussion on #565.
